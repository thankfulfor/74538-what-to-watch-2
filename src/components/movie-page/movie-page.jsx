import * as React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';

import {isObjectEmpty} from '../../utils/is-object-empty.js';

import {getFilmByIdFromUrl, getFilteredFilms} from '../../selector/selectors.js';

import AddToFavoritesButton from '../add-to-favorites-button/add-to-favorites-button.jsx';
import FilmReviews from '../../components/film-reviews/film-reviews.jsx';
import FullScreenVideoPlayer from '../fullscreen-video-player/full-screen-video-player.jsx';
import Header from '../header/header.jsx';
import {FilmDetails} from '../../components/film-details/film-details.jsx';
import {FilmList} from '../film-list/film-list-1.jsx';
import {FilmOverview} from '../../components/film-overview/film-overview.jsx';
import {FilmTabs} from '../film-tabs/film-tabs.jsx';
import {Footer} from '../footer/footer.jsx';
import {PlayButton} from '../play-button/play-button.jsx';

import withOpenCloseButtons from '../../hoc/with-open-close-buttons/with-open-close-buttons.jsx';
import withActiveTab from '../../hoc/with-active-tab/with-active-tab.jsx';

import {history} from '../../history.js';
import {CountConstants, URLS} from '../../utils/constants.js';

const FullscreenVideoPlayerWithButtonsWrapped = withOpenCloseButtons(FullScreenVideoPlayer, PlayButton);
const FilmTabsWrapped = withActiveTab(FilmTabs, FilmDetails, FilmOverview, FilmReviews);

export const MoviePage = (props) => {
  const {film, similarFilms, userData} = props;

  if (isObjectEmpty(film)) {
    return null;
  }

  const {
    background_image: backgroundImage,
    name,
    genre,
    released,
    poster_image: posterImage,
    background_color: backgroundColor,
    id
  } = film;

  const style = {
    backgroundColor: `${backgroundColor}`,
  };

  return (
    <React.Fragment>
      <section className="movie-card movie-card--full" style={style}>
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={backgroundImage} alt={name} />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <Header />

          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{genre}</span>
                <span className="movie-card__year">{released}</span>
              </p>

              <div className="movie-card__buttons">

                <FullscreenVideoPlayerWithButtonsWrapped film={film} />
                <AddToFavoritesButton history={history} filmId={film.id}/>

                {!(Object.entries(userData).length === 0 && userData.constructor === Object) &&
                  <Link to={`${URLS.FILMS_URL}/${id}/review`} className="btn movie-card__button">Add review</Link>
                }
              </div>
            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img src={posterImage} alt={`${name} poster`} width="218" height="327" />
            </div>
            <FilmTabsWrapped film={film} />
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>
          <FilmList films={similarFilms} />
        </section>

        <Footer />
      </div>
    </React.Fragment>
  );
};

MoviePage.propTypes = {
  film: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  similarFilms: PropTypes.array.isRequired,
  userData: PropTypes.object.isRequired,
};

const mapStateToProps = (state, {match}) => {
  const film = getFilmByIdFromUrl(state, match.params.id);
  const similarFilms = getFilteredFilms(state.films, film.genre, CountConstants.COUNT_MORE_LIKE_THIS);
  const userData = state.userData;

  return {film, similarFilms, userData};
};

export default connect(mapStateToProps)(MoviePage);
