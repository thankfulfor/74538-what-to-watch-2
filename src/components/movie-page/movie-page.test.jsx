import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import {BrowserRouter} from 'react-router-dom';
import {MoviePage} from './movie-page.jsx';

Enzyme.configure({adapter: new Adapter()});

const film = {
  "name": `The Revenant`,
  "poster_image": `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/Revenant.jpg`,
  "preview_image": `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/revenant.jpg`,
  "background_image": `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/Revenant.jpg`,
  "background_color": `#92918B`,
  "description": `A frontiersman on a fur trading expedition in the 1820s fights for survival after being mauled by a bear and left for dead by members of his own hunting team.`,
  "rating": 4,
  "scores_count": 618498,
  "director": `Alejandro G. Iñárritu`,
  "starring": [
    `Leonardo DiCaprio`,
    `Tom Hardy`,
    `Will Poulter`
  ],
  "run_time": 156,
  "genre": `Action`,
  "released": 2015,
  "id": 1,
  "is_favorite": false,
  "video_link": `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  "preview_video_link": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
};

it(`MoviePage корректно рендерится после перезапуска`, () => {
  const tree = shallow(
      <BrowserRouter>
        <MoviePage film={film} match={{}} />
      </BrowserRouter>
  );
  expect(toJson(tree)).toMatchSnapshot();
});
