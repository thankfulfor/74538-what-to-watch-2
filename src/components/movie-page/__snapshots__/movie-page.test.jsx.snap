// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePage корректно рендерится после перезапуска 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <MoviePage
    film={
      Object {
        "background_color": "#92918B",
        "background_image": "https://htmlacademy-react-2.appspot.com/wtw/static/film/background/Revenant.jpg",
        "description": "A frontiersman on a fur trading expedition in the 1820s fights for survival after being mauled by a bear and left for dead by members of his own hunting team.",
        "director": "Alejandro G. Iñárritu",
        "genre": "Action",
        "id": 1,
        "is_favorite": false,
        "name": "The Revenant",
        "poster_image": "https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/Revenant.jpg",
        "preview_image": "https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/revenant.jpg",
        "preview_video_link": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
        "rating": 4,
        "released": 2015,
        "run_time": 156,
        "scores_count": 618498,
        "starring": Array [
          "Leonardo DiCaprio",
          "Tom Hardy",
          "Will Poulter",
        ],
        "video_link": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
      }
    }
    match={Object {}}
    similarFilms={Array []}
    userData={Object {}}
  />
</Router>
`;
